crate_universe/extension.bzl

```diff
@@ -0,0 +1,224 @@
+"""Module extension for generating third-party crates for use in bazel."""
+
+load("@bazel_tools//tools/build_defs/repo:git.bzl", "new_git_repository")
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+load("//crate_universe:defs.bzl", _crate_universe_crate = "crate")
+load("//crate_universe/private:crates_vendor.bzl", "CRATES_VENDOR_ATTRS", "generate_config_file", "generate_splicing_manifest")
+load("//crate_universe/private:generate_utils.bzl", "render_config")
+load("//crate_universe/private/module_extensions:cargo_bazel_bootstrap.bzl", "get_cargo_bazel_runner")
+
+def _generate_repo_impl(repo_ctx):
+    for path, contents in repo_ctx.attr.contents.items():
+        repo_ctx.file(path, contents)
+
+_generate_repo = repository_rule(
+    implementation = _generate_repo_impl,
+    attrs = dict(
+        contents = attr.string_dict(mandatory = True),
+    ),
+)
+
+def _generate_annotations(module_ctx, annotation_files):
+    """Given a list of annotation files, merge them into a single json object.
+
+    For an example json file, see examples/bzlmod/hello_world/annotations.json.
+    """
+    annotations = {}
+
+    for file in annotation_files:
+        content = json.decode(module_ctx.read(file))
+        for name, annotations_for_crate in content.items():
+            for annotation in annotations_for_crate:
+                annotation = _crate_universe_crate.annotation(**annotation)
+                annotations.setdefault(name, [])
+                annotations[name].append(annotation)
+
+    return annotations
+
+def _generate_hub_and_spokes(module_ctx, cargo_bazel, mod_path, cfg):
+    annotations = _generate_annotations(module_ctx, cfg.annotation_files)
+    tag_path = mod_path
+    repo_name = "crates"
+    if cfg.suffix:
+        tag_path = mod_path.get_child(cfg.suffix)
+        repo_name += "_" + cfg.suffix
+
+    cargo_lockfile = module_ctx.path(cfg.cargo_lockfile)
+
+    rendering_config = json.decode(render_config(
+        regen_command = "Run 'cargo update [--workspace]'",
```

Comment:
What's the story with this regen_command? I don't think`cargo update` is generally something we should be recommending people run?

Reply:
> I've made it so that the Cargo.lock is the source of truth. The bazel index is based on the Cargo.lock, and is automatically kept up to date.

I'm definitely excited by the idea of being able to get rid of the cargo-bazel lockfile, but I'm pretty sure there were reasons we hadn't already done this... @UebelAndre do you remember?

Some possible things that come to mind, though I'm not 100% confident in any of them:
* The `crate.spec` (via extra crates a versions) and `crate.annotation` (via features) fields can change the resolution, which means if you're not going via our logic you'll miss those parts of the process when you `cargo update`.
* Splicing is expensive (it involves doing a whole resolve, even if it's offline, as well as potentially running `cargo tree` many times to resolve features for platforms) - the cargo-bazel lockfile includes the results of splicing (which is why we can skip it entirely if not repinning), whereas Cargo.lock doesn't so will necessitate re-doing this work. _Possibly_ `Module.bazel.lock` mitigates this as long as it's checked in? But if so, can we get rid of the cargo-bazel lock completely in bzlmod mode?

> There's no need to repin, because repinning can be done via "cargo update".

Even if it's literally just proxying to a call to `cargo update` via the fetched local toolchain, it's important that we support repinning via bazel - we have users who don't themselves have `cargo` installed at all but who may need to trigger a repin.